File: ios/RnNordicMcuFsManagerModule.swift
import ExpoModulesCore
import iOSMcuManagerLibrary

public class RnNordicMcuFsManagerModule: Module, FileDownloadDelegate {
  private var transport: McuMgrBleTransport?
  private var fsManager: FileSystemManager?

  private var jsProgressCallback: JavaScriptFunction<[String: Any?]>?
  private var jsFailedCallback: JavaScriptFunction<[String: Any?]>?
  private var jsCanceledCallback: JavaScriptFunction<[String: Any?]>?
  private var jsCompletedCallback: JavaScriptFunction<[String: Any?]>?

  // Helper to safely call JS callbacks on the JS thread
  private func callCallback(
    _ callback: JavaScriptFunction<[String: Any?]>?,
    payload: [String: Any?]
  ) {
    guard let appContext = self.appContext, let cb = callback else {
      return
    }

    appContext.executeOnJavaScriptThread {
      do {
        _ = try cb.call(payload)
      } catch let err {
        print("RnNordicMcuFsManagerModule: Failed to call JS callback: \(err.localizedDescription)")
      }
    }
  }

  // MARK: - FileDownloadDelegate

  public func downloadProgressDidChange(
    bytesDownloaded: Int,
    fileSize: Int,
    timestamp: Date
  ) {
    print("RnNordicMcuFsManagerModule: Download progress \(bytesDownloaded)/\(fileSize) at \(timestamp)")

    let payload: [String: Any?] = [
      "currentBytes": bytesDownloaded,
      "totalBytes": fileSize,
      // milliseconds since epoch, similar to Android's Long timestamp
      "timestamp": Int(timestamp.timeIntervalSince1970 * 1000)
    ]

    callCallback(jsProgressCallback, payload: payload)
  }

  public func downloadDidFail(with error: Error) {
    print("RnNordicMcuFsManagerModule: Download failed with error: \(error.localizedDescription)")

    let payload: [String: Any?] = [
      "code": "MCU_MGR_DOWNLOAD_FAILED",
      "message": error.localizedDescription,
      "stack": String(describing: error)
    ]

    callCallback(jsFailedCallback, payload: payload)
    cleanupTransport()
  }

  public func downloadDidCancel() {
    print("RnNordicMcuFsManagerModule: Download was canceled")

    let payload: [String: Any?] = [
      "canceled": true
    ]

    callCallback(jsCanceledCallback, payload: payload)
    cleanupTransport()
  }

  public func download(of name: String, didFinish data: Data) {
    print("RnNordicMcuFsManagerModule: Download of \(name) finished successfully, bytes=\(data.count)")

    // Data -> [Int] (0–255) to match Android’s List<Int>
    let bytes: [Int] = data.map { Int($0) & 0xFF }

    let payload: [String: Any?] = [
      "data": bytes,
      "size": data.count
    ]

    callCallback(jsCompletedCallback, payload: payload)
    cleanupTransport()
  }

  private func cleanupTransport() {
    // Close filesystem manager & BLE transport
    fsManager = nil
    transport?.close()
    transport = nil

    jsProgressCallback = nil
    jsFailedCallback = nil
    jsCanceledCallback = nil
    jsCompletedCallback = nil
  }

  // MARK: - Module definition

  public func definition() -> ModuleDefinition {
    Name("RnNordicMcuFsManager")

    /**
     * destroy() – stateless implementation, just ensures we close any active transport.
     */
    Function("destroy") {
      print("RnNordicMcuFsManagerModule: destroy() called")
      self.cleanupTransport()
    }

    /**
     * fileDownload(
     *   deviceId: string,
     *   filename: string,
     *   onDownloadProgressChanged?: (progress: { currentBytes, totalBytes, timestamp }) => void,
     *   onDownloadFailed?: (error: { code, message, stack? }) => void,
     *   onDownloadCanceled?: (info: { canceled: true }) => void,
     *   onDownloadCompleted?: (result: { data: number[], size: number }) => void
     * )
     */
    Function("fileDownload") {
      (
        deviceId: String,
        filename: String,
        progressCallback: JavaScriptFunction<[String: Any?]>?,
        failedCallback: JavaScriptFunction<[String: Any?]>?,
        canceledCallback: JavaScriptFunction<[String: Any?]>?,
        completedCallback: JavaScriptFunction<[String: Any?]>?
      ) in

      print("RnNordicMcuFsManagerModule: fileDownload(deviceId=\(deviceId), filename=\(filename))")

      // Store callbacks
      self.jsProgressCallback = progressCallback
      self.jsFailedCallback = failedCallback
      self.jsCanceledCallback = canceledCallback
      self.jsCompletedCallback = completedCallback

      // Init transport + FS manager for this download (stateless like Android)
      guard let uuid = UUID(uuidString: deviceId) else {
        let msg = "Failed to parse UUID from deviceId: \(deviceId)"
        print("RnNordicMcuFsManagerModule: \(msg)")

        let payload: [String: Any?] = [
          "code": "INIT_ERROR",
          "message": msg,
          "stack": msg
        ]
        self.callCallback(failedCallback, payload: payload)
        return
      }

      do {
        self.transport = McuMgrBleTransport(uuid)
        guard let transport = self.transport else {
          let msg = "Failed to create McuMgrBleTransport"
          print("RnNordicMcuFsManagerModule: \(msg)")
          let payload: [String: Any?] = [
            "code": "INIT_ERROR",
            "message": msg,
            "stack": msg
          ]
          self.callCallback(failedCallback, payload: payload)
          return
        }

        self.fsManager = FileSystemManager(transport: transport)

        guard let fsManager = self.fsManager else {
          let msg = "Failed to create FileSystemManager"
          print("RnNordicMcuFsManagerModule: \(msg)")
          let payload: [String: Any?] = [
            "code": "INIT_ERROR",
            "message": msg,
            "stack": msg
          ]
          self.callCallback(failedCallback, payload: payload)
          self.cleanupTransport()
          return
        }

        _ = fsManager.download(name: filename, delegate: self)
      } catch let err {
        print("RnNordicMcuFsManagerModule: Error starting file download: \(err.localizedDescription)")
        let payload: [String: Any?] = [
          "code": "START_DOWNLOAD_ERROR",
          "message": err.localizedDescription,
          "stack": String(describing: err)
        ]
        self.callCallback(failedCallback, payload: payload)
        self.cleanupTransport()
      }
    }
  }
}

File: android/src/main/java/com/codedlines/rnnordicmcufsmanager/RnNordicMcuFsManagerModule.kt
package com.codedlines.rnnordicmcufsmanager

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.content.Context
import android.util.Log
import expo.modules.kotlin.jni.JavaScriptFunction
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import io.runtime.mcumgr.McuMgrTransport
import io.runtime.mcumgr.ble.McuMgrBleTransport
import io.runtime.mcumgr.exception.McuMgrException
import io.runtime.mcumgr.managers.FsManager
import io.runtime.mcumgr.transfer.DownloadCallback

private const val TAG = "RnNordicMcuFsManagerModule"

class RnNordicMcuFsManagerModule : Module() {

    private val context: Context
        get() = appContext.reactContext
            ?: appContext.currentActivity
            ?: throw IllegalStateException("React or Activity context is null")

    private fun getBluetoothDevice(macAddress: String): BluetoothDevice {
        val bluetoothManager =
            context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
                ?: throw IllegalStateException("BluetoothManager not available")

        val adapter = bluetoothManager.adapter
            ?: throw IllegalStateException("No Bluetooth adapter available")

        return try {
            adapter.getRemoteDevice(macAddress)
        } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid MAC / device id: $macAddress", e)
        }
    }

    private fun safeInvokeCallback(
        callback: JavaScriptFunction<Any?>?,
        payload: Any? = null
    ) {
        if (callback == null) return

        appContext.executeOnJavaScriptThread {
            try {
                if (payload != null) {
                    callback.invoke(payload)
                } else {
                    callback.invoke()
                }
            } catch (t: Throwable) {
                Log.e(TAG, "Error while invoking JS callback", t)
            }
        }
    }

    override fun definition() = ModuleDefinition {
        Name("RnNordicMcuFsManager")

        /**
         * destroy() – no-op for stateless implementation, kept for API compatibility.
         */
        Function("destroy") {
            Log.d(TAG, "destroy() called – stateless implementation, nothing to clean up")
        }

        /**
         * fileDownload(
         *   deviceId: string,
         *   filename: string,
         *   onDownloadProgressChanged?: (progress: { currentBytes, totalBytes, timestamp }) => void,
         *   onDownloadFailed?: (error: { code, message, stack? }) => void,
         *   onDownloadCanceled?: (info: { canceled: true }) => void,
         *   onDownloadCompleted?: (result: { data: number[], size: number }) => void
         * )
         */
        Function("fileDownload") {
                macAddress: String,
                filename: String,
                onDownloadProgressChanged: JavaScriptFunction<Any?>?,
                onDownloadFailed: JavaScriptFunction<Any?>?,
                onDownloadCanceled: JavaScriptFunction<Any?>?,
                onDownloadCompleted: JavaScriptFunction<Any?>?
            ->

            Log.d(TAG, "fileDownload() called with mac=$macAddress, filename=$filename")

            val device: BluetoothDevice
            val transport = McuMgrBleTransport(context, device).apply {
                setInitialMtu(498)            // default, but make it explicit
                setMaxPacketLength(498)       // allow full-size SMP frames
                requestConnPriority(BluetoothGatt.CONNECTION_PRIORITY_HIGH)
            }
            val manager: FsManager

            try {
                device = getBluetoothDevice(macAddress)
                transport = McuMgrBleTransport(context, device)
                manager = FsManager(transport)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to create transport / FsManager", e)
                val payload = mapOf(
                    "code" to "INIT_ERROR",
                    "message" to (e.message ?: "Failed to init transport or FsManager"),
                    "stack" to e.stackTraceToString()
                )
                safeInvokeCallback(onDownloadFailed, payload)
                throw e
            }

            fun releaseTransport() {
                try {
                    (transport as? McuMgrBleTransport)?.release()
                } catch (e: Exception) {
                    Log.e(TAG, "Error releasing transport", e)
                }
            }

            val downloadCallback = object : DownloadCallback {
                override fun onDownloadProgressChanged(
                    current: Int,
                    total: Int,
                    timestamp: Long
                ) {
                    Log.d(
                        TAG,
                        "Download progress: current=$current, total=$total, timestamp=$timestamp"
                    )

                    val payload = mapOf(
                        "currentBytes" to current,
                        "totalBytes" to total,
                        "timestamp" to timestamp
                    )

                    safeInvokeCallback(onDownloadProgressChanged, payload)
                }

                override fun onDownloadFailed(e: McuMgrException) {
                    Log.e(TAG, "Download failed", e)

                    val payload = mapOf(
                        "code" to "MCU_MGR_DOWNLOAD_FAILED",
                        "message" to (e.message ?: "Unknown McuMgrException"),
                        "stack" to e.stackTraceToString()
                    )

                    safeInvokeCallback(onDownloadFailed, payload)
                    releaseTransport()
                }

                override fun onDownloadCanceled() {
                    Log.w(TAG, "Download canceled")

                    val payload = mapOf(
                        "canceled" to true
                    )

                    safeInvokeCallback(onDownloadCanceled, payload)
                    releaseTransport()
                }

                override fun onDownloadCompleted(data: ByteArray) {
                    Log.d(TAG, "Download completed, bytes=${data.size}")

                    val bytes: List<Int> = data.map { it.toInt() and 0xFF }

                    val payload = mapOf(
                        "data" to bytes,
                        "size" to data.size
                    )

                    safeInvokeCallback(onDownloadCompleted, payload)
                    releaseTransport()
                }
            }

            try {
                Log.d(TAG, "Starting file download over mcumgr")
                manager.fileDownload(filename, downloadCallback)
            } catch (e: Exception) {
                Log.e(TAG, "Error starting file download for: $filename", e)

                val payload = mapOf(
                    "code" to "START_DOWNLOAD_ERROR",
                    "message" to (e.message ?: "Unknown error starting download"),
                    "stack" to e.stackTraceToString()
                )

                safeInvokeCallback(onDownloadFailed, payload)
                releaseTransport()
                throw e
            }
        }
    }
}
